Index: b/arch/mips/ralink/mt7620.c
===================================================================
--- a/arch/mips/ralink/mt7620.c	2017-09-12 10:20:01.956341626 +0800
+++ b/arch/mips/ralink/mt7620.c	2017-09-12 10:24:57.206066813 +0800
@@ -138,8 +138,13 @@ static struct rt2880_pmx_func i2c_grp_mt
 	FUNC("i2c", 0, 4, 2),
 };
 
-static struct rt2880_pmx_func refclk_grp_mt7628[] = { FUNC("reclk", 0, 36, 1) };
-static struct rt2880_pmx_func perst_grp_mt7628[] = { FUNC("perst", 0, 37, 1) };
+static struct rt2880_pmx_func refclk_grp_mt7628[] = { 
+	FUNC("gpio", 1, 37, 1),
+	FUNC("refclk", 0, 37, 1),
+};
+
+//static struct rt2880_pmx_func refclk_grp_mt7628[] = { FUNC("reclk", 0, 36, 1) };
+static struct rt2880_pmx_func perst_grp_mt7628[] = { FUNC("perst", 0, 36, 1) };
 static struct rt2880_pmx_func wdt_grp_mt7628[] = { FUNC("wdt", 0, 38, 1) };
 static struct rt2880_pmx_func spi_grp_mt7628[] = { FUNC("spi", 0, 7, 4) };
 
@@ -172,7 +177,7 @@ static struct rt2880_pmx_func spi_cs1_gr
 };
 
 static struct rt2880_pmx_func spis_grp_mt7628[] = {
-	FUNC("pwm", 3, 14, 4),
+	FUNC("pwm_uart2", 3, 14, 4),
 	FUNC("util", 2, 14, 4),
 	FUNC("gpio", 1, 14, 4),
 	FUNC("spis", 0, 14, 4),
@@ -180,7 +185,7 @@ static struct rt2880_pmx_func spis_grp_m
 
 static struct rt2880_pmx_func gpio_grp_mt7628[] = {
 	FUNC("pcie", 3, 11, 1),
-	FUNC("refclk", 2, 11, 1),
+	FUNC("refclk_fc", 2, 11, 1),
 	FUNC("gpio", 1, 11, 1),
 	FUNC("gpio", 0, 11, 1),
 };
@@ -193,16 +198,64 @@ static struct rt2880_pmx_func wled_kn_gr
 };
 
 static struct rt2880_pmx_func wled_an_grp_mt7628[] = {
-	FUNC("rsvd", 3, 35, 1),
-	FUNC("rsvd", 2, 35, 1),
-	FUNC("gpio", 1, 35, 1),
-	FUNC("wled_an", 0, 35, 1),
+	FUNC("rsvd", 3, 44, 1),
+	FUNC("rsvd", 2, 44, 1),
+	FUNC("gpio", 1, 44, 1),
+	FUNC("wled_an", 0, 44, 1),
+};
+
+static struct rt2880_pmx_func p0_led_an_grp_mt7628[] = {
+        FUNC("jtag", 3, 43, 1),
+        FUNC("rsvd", 2, 43, 1),
+        FUNC("gpio", 1, 43, 1),
+        FUNC("ephy", 0, 43, 1),
+};
+
+static struct rt2880_pmx_func p1_led_an_grp_mt7628[] = {
+        FUNC("jtag", 3, 42, 1),
+        FUNC("utif", 2, 42, 1),
+        FUNC("gpio", 1, 42, 1),
+        FUNC("ephy", 0, 42, 1),
+};
+
+static struct rt2880_pmx_func p2_led_an_grp_mt7628[] = {
+        FUNC("jtag", 3, 41, 1),
+        FUNC("utif", 2, 41, 1),
+        FUNC("gpio", 1, 41, 1),
+        FUNC("ephy", 0, 41, 1),
+};
+
+
+static struct rt2880_pmx_func p3_led_an_grp_mt7628[] = {
+        FUNC("jtag", 3, 40, 1),
+        FUNC("utif", 2, 40, 1),
+        FUNC("gpio", 1, 40, 1),
+        FUNC("ephy", 0, 40, 1),
+};
+
+static struct rt2880_pmx_func p4_led_an_grp_mt7628[] = {
+        FUNC("jtag", 3, 39, 1),
+        FUNC("utif", 2, 39, 1),
+        FUNC("gpio", 1, 39, 1),
+        FUNC("ephy", 0, 39, 1),
 };
 
 #define MT7628_GPIO_MODE_MASK		0x3
 
+#define MT7628_GPIO_MODE_P4_LED_KN	58
+#define MT7628_GPIO_MODE_P3_LED_KN	56
+#define MT7628_GPIO_MODE_P2_LED_KN	54
+#define MT7628_GPIO_MODE_P1_LED_KN	52
+#define MT7628_GPIO_MODE_P0_LED_KN	50
 #define MT7628_GPIO_MODE_WLED_KN	48
+
+#define MT7628_GPIO_MODE_P4_LED_AN	42
+#define MT7628_GPIO_MODE_P3_LED_AN	40
+#define MT7628_GPIO_MODE_P2_LED_AN	38
+#define MT7628_GPIO_MODE_P1_LED_AN	36
+#define MT7628_GPIO_MODE_P0_LED_AN	34
 #define MT7628_GPIO_MODE_WLED_AN	32
+
 #define MT7628_GPIO_MODE_PWM1		30
 #define MT7628_GPIO_MODE_PWM0		28
 #define MT7628_GPIO_MODE_UART2		26
@@ -236,7 +289,12 @@ static struct rt2880_pmx_group mt7628an_
 	GRP_G("spis", spis_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_SPIS),
 	GRP_G("gpio", gpio_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_GPIO),
 	GRP_G("wled_an", wled_an_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_WLED_AN),
-	GRP_G("wled_kn", wled_kn_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_WLED_KN),
+	//GRP_G("wled_kn", wled_kn_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_WLED_KN),
+	GRP_G("p0_led_an", p0_led_an_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_P0_LED_AN),
+	GRP_G("p1_led_an", p1_led_an_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_P1_LED_AN),
+	GRP_G("p2_led_an", p2_led_an_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_P2_LED_AN),
+	GRP_G("p3_led_an", p3_led_an_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_P3_LED_AN),
+	GRP_G("p4_led_an", p4_led_an_grp_mt7628, MT7628_GPIO_MODE_MASK, 1, MT7628_GPIO_MODE_P4_LED_AN),
 	{ 0 }
 };
 
